### PR TITLE
Add support for custom delimiter in line-based importer

### LINKED ISSUE #4103: Allow line-based importer to specify a custom line-delimiter sequence for custom data
This might need better streaming support for our line-based importer as a pre-requisite.  I don't know.
I often land myself with large byte arrays streamed out to a single file that have custom sequence of chars (`*%%*`) used as record delimiters that I would like to treat simply as a line-delimiter while importing the file into OpenRefine.  The file sizes are typically under 4GB, usually only 1GB or 2GB in size, where I often have over 20GB system memory available to give to the Java heap.
Other tools allow reading a file as a stream of characters and separating into new lines based on a custom char sequence.
I would like our OpenRefine line-based importer or a new importer to handle this use case.

### Proposed solution
Allow line-based importer to have an option to use a custom line-delimiter character sequence (overriding the defaults of `\n` or `\r\n`.

### Alternatives considered
I have to use other tools.

### PR DESCRIPTION
**Pull request by EgorDm**
Fixes #4103

Changes proposed in this pull request:
- LineBasedImporter now uses Scanner with regex delimiter instead of a LineNumberReader
- Updated UI to select between newline splitting (default) and regex-based splitting.
- Added tests for basic LineBasedImporter functionality and regex splitting

Note: Scanner also does buffering so adding BufferedReader in between is not necessary. Though the buffer is smaller.

### CODE REVIEW

**Review by wetneb — APPROVED** [review]
Amazing, thank you so much @EgorDm!
I tested it and it works flawlessly. The code and tests look perfect too.
Thumbs up on my side! Merging soon unless anyone has concerns.

File: main/tests/server/src/com/google/refine/importers/LineBasedImporterTests.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+import org.slf4j.LoggerFactory;
+import org.testng.Assert;
+import org.testng.annotations.*;
+
+import java.io.Reader;
+import java.io.StringReader;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.testng.Assert.assertEquals;
```
Comment (tfmorris) [inline main/tests/server/src/com/google/refine/importers/LineBasedImporterTests.java:46]: This static import is never used. All the `Assert.assertEquals` mentions can be changed to just `assertEquals`

File: main/tests/server/src/com/google/refine/importers/LineBasedImporterTests.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY           
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+package com.google.refine.importers;
+
+import com.google.refine.util.ParsingUtilities;
```
Comment (tfmorris) [inline main/tests/server/src/com/google/refine/importers/LineBasedImporterTests.java:36]: This import is unused. It and the other unused imports should be removed.

**Review by tfmorris — CHANGES_REQUESTED** [review]
Thanks for the contribution! Just a few cosmetic issues to clean up.

File: main/src/com/google/refine/importers/LineBasedImporter.java [hunk 3]
```diff
@@ -87,15 +100,16 @@ public void parseOneFile(
...
 
         try {
             int skip = JSONUtilities.getInt(options, "ignoreLines", -1);
             while (skip > 0) {
-                lnReader.readLine();
+                lnReader.next();
                 skip--;
             }
-        } catch (IOException e) {
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (tfmorris) [reply main/src/com/google/refine/importers/LineBasedImporter.java:112]: Actually, on further investigation, this code doesn't throw any exceptions at all! The entire try/catch block should be removed.

[the remark this answers is not in the document]
Comment (EgorDm) [reply main/src/com/google/refine/importers/LineBasedImporter.java]: Thanks for the review.

Good call. 
Scanner seems to catch IO errors and put them in a member variable. 
I have removed try catch and added the following block.

```
if (lnReader.ioException() != null) {
            logger.error("Error reading line-based file", lnReader.ioException());
        }
```

File: main/tests/server/src/com/google/refine/importers/LineBasedImporterTests.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
+/*
+
+Copyright 2010,2011 Google Inc.
```
[the remark this answers is not in the document]
Comment (EgorDm) [reply main/tests/server/src/com/google/refine/importers/LineBasedImporterTests.java]: Should be fixed in the latest commit.

**Review by tfmorris — APPROVED** [review]
LGTM Thanks for the updates!

**Pull request merged**